### PR TITLE
[RemoveDIs] Update Clang front end to handle DbgRecords

### PR DESCRIPTION
**Pull request by OCHyams**
This patch fixes problems that pop up when clang emits DbgRecords instead of debug intrinsics.

Note: this doesn't mean clang is emitting DbgRecords yet, because the modules it creates are still always in the old debug mode. That will come in a future patch.

Depends on #84739

### CODE REVIEW

File: clang/lib/CodeGen/CGStmtOpenMP.cpp [hunk 1]
```diff
@@ -4746,20 +4746,32 @@ void CodeGenFunction::EmitOMPTaskBasedDirective(
           if (CGF.CGM.getCodeGenOpts().hasReducedDebugInfo())
             (void)DI->EmitDeclareOfAutoVariable(SharedVar, ContextValue,
                                                 CGF.Builder, false);
-          llvm::Instruction &Last = CGF.Builder.GetInsertBlock()->back();
           // Get the call dbg.declare instruction we just created and update
           // its DIExpression to add offset to base address.
-          if (auto DDI = dyn_cast<llvm::DbgVariableIntrinsic>(&Last)) {
+          auto UpdateExpr = [](llvm::LLVMContext &Ctx, auto *DDI,
```
Comment (SLTozer) [inline clang/lib/CodeGen/CGStmtOpenMP.cpp:4751]: Super nit, but in situations like this I prefer to change things like `DDI` to `Declare`; subjective, so non-blocking.

**Review by SLTozer — COMMENTED** [review]
Maybe wants someone with more frontend knowledge to chime in, but all SGTM.

**Comment by OCHyams** [comment]
> Maybe wants someone with more frontend knowledge to chime in, but all SGTM.

@adrian-prantl are you able to suggest anyone that might be able to take a look at this (part of this touches some Objective-C code+test) - we're not overly familiar with front end code.

**Review by SLTozer — APPROVED** [review]
It's been a bit without any comments, probably safe now to land this.

**Pull request merged**